### PR TITLE
feat(freespace_planning_algorithms): implement option for backward search from goal to start

### PR DESCRIPTION
**Pull request by mkquda**
## Description
When using A-star freespace planner, in most cases the path found will arrive at the goal with small offset in position and orientation due to the discretized search space. As a result the vehicle might not arrive with the exact desired pose.

For certain applications such as parking, matching the actual goal pose is important, therefore it useful to be able to configure the planner to do the search process backwards (starting from goal pose towards start pose).

## Changes
- Includes changes from preceding [PR](https://github.com/autowarefoundation/autoware.universe/pull/8089)
- Modified A-star expansion logic to support backward search
- Added parameter to configure search method (forward search or backward search)

## Related links

**Preceding PR's**
- https://github.com/autowarefoundation/autoware.universe/pull/8068
- https://github.com/autowarefoundation/autoware.universe/pull/8089

## How was this PR tested?
tested on psim.
compared performance with unit tests:

### Single Turning Step
(config: max_turning_ratio = 0.5, grid_resolution = 0.3, theta_size = 120)
Screenshot from 2024-07-22 18-17-03

### Multi Turning Steps
(config: max_turning_ratio = 0.5, grid_resolution = 0.3, theta_size = 120)
Screenshot from 2024-07-22 18-17-35

## Notes for reviewers

This PR includes changes from a preceding [PR](https://github.com/autowarefoundation/autoware.universe/pull/8089). so please review the PR's in order.

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added | `search_method`   | `string` | `forward` | method of searching, start to goal or vice versa |

## Effects on system behavior

Can configure A-star search direction, forward search (start pose to goal pose) OR backward search (goal pose to start pose)

### CODE REVIEW

**Review by armaganarsln — CHANGES_REQUESTED** [review]
Can you please test with this issue: 
- https://github.com/autowarefoundation/autoware.universe/issues/8342

**Review by armaganarsln — APPROVED** [review]

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 1]
```diff
@@ -333,46 +333,60 @@ void AstarSearch::setPath(const AstarNode & goal_node)
...
+  for (; it < waypoints.end() - 1; ++it) {
+    auto next_it = it + 1;
+    waypoints_.waypoints.push_back(*it);
+    if (it->is_back == next_it->is_back) continue;
+    if (is_backward_search_) {
+      waypoints_.waypoints.push_back({next_it->pose, it->is_back});
+    } else {
+      waypoints_.waypoints.push_back({it->pose, next_it->is_back});
+    }
   }
```
Comment (xmfcx) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:388]:
```suggestion
  waypoints_.waypoints.reserve(waypoints.size());
  
  for (size_t i = 0; i < waypoints.size() - 1; ++i) {
    const auto& current = waypoints[i];
    const auto& next = waypoints[i + 1];

    waypoints_.waypoints.push_back(current);

    if (current.is_back != next.is_back) {
      waypoints_.waypoints.push_back({
        is_backward_search_ ? next.pose : current.pose,
        is_backward_search_ ? current.is_back : next.is_back
      });
    }
  }
```

Maybe something like this to keep it more readable.

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 2]
```diff
@@ -333,46 +333,60 @@ void AstarSearch::setPath(const AstarNode & goal_node)
...
 
   // push astar nodes poses
   while (node != nullptr) {
     pose.pose = local2global(costmap_, node2pose(*node));
-    waypoints_.waypoints.push_back({pose, node->is_back});
+    waypoints.push_back({pose, node->is_back});
     interpolate(*node);
     // To the next node
     node = node->parent;
   }
```
Comment (xmfcx) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:368]: Maybe add some check to make sure `waypoints` is not empty here. Or even to make sure it has at least 2 elements, not sure. Some operations below here depend on this.

**Comment by xmfcx** [comment]
Could you rebase to the latest main?

**Review by maxime-clem — APPROVED** [review]
Looks good and parking works fine in Psim.

**Review by kosuke55 — APPROVED** [review]
LGTM for behavior path planner

**Pull request merged**